apps/aecore/lib/aecore/oracle/oracle_state_tree.ex

```diff
@@ -189,12 +191,49 @@ defmodule Aecore.Oracle.OracleStateTree do
     end
   end
 
-  defp which_tree(tree, :oracle), do: tree.oracle_tree
-  defp which_tree(tree, :oracle_query), do: tree.oracle_tree
-  defp which_tree(tree, _where), do: tree.oracle_tree
+  defp which_tree(oracles_state, :oracle), do: oracles_state.oracle_tree
+  defp which_tree(oracles_state, :oracle_query), do: oracles_state.oracle_tree
+  defp which_tree(oracles_state, _where), do: oracles_state.oracle_tree
 
   defp cache_push(oracle_cache_tree, key, expires) do
     encoded = Serialization.cache_key_encode(key, expires)
     enter(oracle_cache_tree, encoded, @dummy_val)
   end
+
+  defp init_expired_cache_key_removal(oracles_state) do
+    %{oracle_cache_tree: cache_tree} =
+      oracles_state.oracle_cache_tree
+      |> PatriciaMerkleTree.all_keys()
+      |> Enum.reduce(oracles_state, fn key, new_state ->
+        new_cache_tree =
+          key
+          |> Serialization.cache_key_decode()
+          |> remove_expired_cache_key(key, new_state)
+
+        %{new_state | oracle_cache_tree: new_cache_tree}
+      end)
+
+    cache_tree
+  end
+
+  defp remove_expired_cache_key({exp, {:oracle, id}}, expired_cache_key, oracles_state) do
+    oracle = get(oracles_state.oracle_tree, id)
+
+    if oracle.expires > exp do
+      delete(oracles_state.oracle_cache_tree, expired_cache_key)
+    else
+      oracles_state.oracle_cache_tree
+    end
+  end
+
+  defp remove_expired_cache_key({exp, {:query, oracle_id, id}}, expired_cache_key, oracles_state) do
+    query_id = oracle_id <> id
```

Comment:
This is the only row added to the previous function and the other code is copied here. Maybe the repeated code could be extracted to another function, but this should not be leaved as is.